### PR TITLE
banana-i18n integration with simulator

### PR DESCRIPTION
**Pull request by pavanjoshi914**
### quick glance of the library
1. define constructor eg: ```banana```
2. access all the methods provided by library via ```banana.method_name()```
3. set locale value using ```setLocale```
4. setup ```fetch api``` accordingly to load translations
5. use methods in js files

### more  Setup information
1. ```banana-i18n.js``` library file```
2. ```config-i18n.js``` configuration for i18n.
3. translation loading is done using ```fetch API``` with ```async ```and``` await``` so that function is not completed till fetch request is completed.
4. cache is disabled during ```http request``` so that when ```json``` is modified ```json.response()``` will not contain old data and will render new data
5. locale is brought from backend to frontend using ```window.locale```
6. other functionalities such as plurals , interpolations/placeholders working fine 

### things needed to be handled further
1. page should not load until translations are loaded via fetch api.
sometimes program executes and delay in fetch can result in printing of key names only and not actual translations.
2. error handling 
fetch api doesnt handle 404 request. logic has to be build for handiling server and bad http request errors such as 404 file not found.
3. library is not resolving fallback language as well as fallback messages with webacker . (work in progress)

### Screenshots of the changes (If any) -

Note: Please check **Allow edits from maintainers.** if you would like us to assist in the PR.

### CODE REVIEW

**Comment by pavanjoshi914** [comment]
CC @satu0king @tachyons

File: public/js/config-i18n.js [hunk 1]
```diff
@@ -0,0 +1,19 @@
+// intialize banana-i18n constructor
+const banana = new Banana({ finalFallback: 'en' });
+// set locale
+banana.setLocale(window.locale);
+// setup fetch api with async/await to load json files
+async function fetchJson() {
```
Comment (tachyons) [inline public/js/config-i18n.js:6]: Can we fetch the locales directly instead of making a http call ?

Comment (pavanjoshi914) [reply public/js/config-i18n.js:6]: Thats seems the drawback of the library they only provide ```banana.load()``` which can hold the constant having messages defined inside it . for loading messages from json explicitly they suggest fetch()

here is the documentation - https://github.com/wikimedia/banana-i18n#loading-the-messages
there demo example uses the same  - https://github.com/wikimedia/banana-i18n/blob/master/demo/js/demo.js

Current library is a model of``` jquery.i18n ```and purely written in js .  Jquery I18n provides load method to load translations but not this one
https://github.com/wikimedia/jquery.i18n#dynamic-loading-using-load-method

Comment (tachyons) [reply public/js/config-i18n.js:6]: You can load the i18n translations to messages object and pass it to the constructor. It is just that banana is doesn't dictate a directory structure. You need to implement loading yourself

Comment (pavanjoshi914) [reply public/js/config-i18n.js:6]: Yes, loading is to be done manually, so gone with fetch API 

Now I am thinking of bringing translations in form of J's objects and parse them using json.parse for  further use

Any suitable loading technique you want to suggest here? As there can be more than one techniques?

Comment (tachyons) [reply public/js/config-i18n.js:6]: Using fetch require http call, and it won't play nicely with the offline app. Here is one approach with load method https://github.com/zhaofengli/refill/blob/2ee4c00623f20bbf9b86e76679f763217f7a241e/web/src/main.js#L36

Comment (tachyons) [reply public/js/config-i18n.js:6]: Another approach https://github.com/smc/mlmorph-web/blob/5d741f617c0ed541fb65a249b039d74d1daf4d15/src/main.js#L9

File: public/i18n/en.json [hunk 2]
```diff
@@ -0,0 +1,12 @@
+{
```
Comment (tachyons) [inline public/i18n/en.json:1]: this  should be in simulator folder

Comment (pavanjoshi914) [reply public/i18n/en.json:1]: yes , desired the same but fetch api makes requests such as ```localhost/i18n/en.json```  in that case had to use public directory .

will modify this also as i change loading mechanism

File: app/views/layouts/simulator.html.erb [hunk 3]
```diff
@@ -3,7 +3,11 @@
   <head>
     <title>CircuitVerse - Digital Circuit Simulator online</title>
     <%= csrf_meta_tags %>
+    <!-- bring locale value from backend to frontend -->
+    <script>window.locale = "<%= I18n.locale %>"</script>
     <script src="/js/pace.js"></script>
+    <!--load banana-i18n library-->
+   <script src="/js/banana-i18n.js"></script>
```
Comment (tachyons) [inline app/views/layouts/simulator.html.erb:10]: Do you still need this as we are fetching from npm?

Comment (pavanjoshi914) [reply app/views/layouts/simulator.html.erb]: I did mistake here, now importing from node_modules. corrected in new commit!

File: simulator/src/combinationalAnalysis.js [hunk 4]
```diff
@@ -11,6 +11,14 @@ import AndGate from './modules/AndGate';
 import OrGate from './modules/OrGate';
 import NotGate from './modules/NotGate';
 import { stripTags } from './utils';
+import banana from './config-i18n';
+
+
```
Comment (tachyons) [inline simulator/src/combinationalAnalysis.js:16]:
```suggestion

```

File: simulator/src/combinationalAnalysis.js [hunk 5]
```diff
@@ -11,6 +11,12 @@ import AndGate from './modules/AndGate';
 import OrGate from './modules/OrGate';
 import NotGate from './modules/NotGate';
 import { stripTags } from './utils';
+import banana from './config-i18n';
+// library basic test
+console.log(banana.i18n('appname-title'));
+// interpolation and plurals check
+console.log(banana.i18n('result', 1));
+console.log(banana.i18n('result', 4));
```
Comment (tachyons) [inline simulator/src/combinationalAnalysis.js:19]:
```suggestion
```

Comment (tachyons) [reply simulator/src/combinationalAnalysis.js:19]: Remove console log

File: simulator/src/combinationalAnalysis.js [hunk 6]
```diff
@@ -26,7 +27,7 @@ var sampleOutputListNames = ['X'];
 export function createCombinationalAnalysisPrompt(scope = globalScope) {
     scheduleBackup();
     $('#combinationalAnalysis').empty();
-    $('#combinationalAnalysis').append("<p>Enter Input names separated by commas: <input id='inputNameList' type='text'  placeHolder='eg. In A, In B'></p>");
+    $('#combinationalAnalysis').append(banana.i18n('appname-title'));
```
Comment (tachyons) [inline simulator/src/combinationalAnalysis.js:30]: You need proper translation here

**Comment by pavanjoshi914** [comment]
I was thinking of creating fresh PR with single clean commit...
If discussion is required for future reference I will continue with this.

**Comment by tachyons** [comment]
Feel free to create New PR

**Comment by pavanjoshi914** [comment]
closing this , new pr for the same -  https://github.com/CircuitVerse/CircuitVerse/pull/2318

**Pull request closed without merging**